bionic/core/task_execution.py

```diff
@@ -478,7 +478,13 @@ def set_up_caching_flags(self, bootstrap):
             should_memoize = True
         self.should_memoize = should_memoize
 
-        should_persist = can_persist and not self.output_would_be_missing()
+        should_persist = optional_should_persist
+        if should_persist is None:
+            if bootstrap is None:
+                should_persist = False
+            else:
+                should_persist = bootstrap.should_persist_default
+        should_persist = should_persist and not self.output_would_be_missing()
```

Comment:
WDYT about writing it like this?

    if self.output_would_be_missing():
        should_persist = False
    elif optional_should_persist is not None:
        should_persist = optional_should_persist
    elif bootstrap is not None:
        should_persist = bootstrap.should_persist_default
    else:
        should_persist = False

This avoids mutation and feels easier for me to follow than the current code (which sets `should_persist`, then reads it, then sets it, then reads it, then sets it).